go/libraries/doltcore/table/editor/pk_table_editor.go

```diff
@@ -276,14 +334,27 @@ func (te *pkTableEditor) InsertRow(ctx context.Context, dRow row.Row) error {
 	te.writeMutex.Lock()
 	defer te.writeMutex.Unlock()
 
-	// If we've already inserted this key as part of this insert operation, that's an error. Inserting a row that
-	// already exists in the table will be handled in Close().
+	// If we've already inserted this key as part of this batch then return an error.
 	if _, ok := te.tea.addedKeys[keyHash]; ok {
 		keyStr, err := formatKey(ctx, key)
 		if err != nil {
 			return err
 		}
-		return fmt.Errorf(ErrDuplicatePrimaryKeyFmt, keyStr)
+		return sql.ErrUniqueKeyViolation.New(keyStr)
+	}
+	// Check if the key is in the existing data, and is not scheduled to be removed.
+	if _, ok := te.tea.removedKeys[keyHash]; !ok {
+		pkExists, err := te.rowData.Has(ctx, keyHash, key)
+		if err != nil {
+			return errhand.BuildDError("failed to read table").AddCause(err).Build()
```

Comment:
Only simple errors in this package.

Reply:
This was actually already in this package. We actually use it quite a bit in this same file in other places. I can remove those though.